backend/core/src/listings/listings.service.ts

```diff
@@ -49,62 +49,77 @@ export class ListingsService {
-        qb
+        innerFilteredQuery,
+        innerWhereParams
       )
     }
 
     qb.orderBy({
       "listings.id": "DESC",
+      "units.max_occupancy": "ASC",
+      "preferences.ordinal": "ASC",
     })
 
-    let currentPage: number = params.page
-    let itemsPerPage: number = params.limit
-
-    let itemCount: number, totalItemsCount: number, totalPages: number
+    const paginationInfo = {
+      currentPage: params.page,
+      itemCount: undefined as number,
+      itemsPerPage: params.limit,
+      totalItems: undefined as number,
+      totalPages: undefined as number,
+    }
     let listings: Listing[]
 
-    if (currentPage > 0 && itemsPerPage > 0) {
-      // Calculate the number of listings to skip (because they belong to lower page numbers)
-      const skip = (currentPage - 1) * itemsPerPage
-      qb = qb.skip(skip).take(itemsPerPage)
-
-      listings = await qb.getMany()
-
-      itemCount = listings.length
-
-      // Issue a separate "COUNT(*) from listings;" query to get the total listings count.
-      totalItemsCount = await this.listingRepository.count()
-      totalPages = Math.ceil(totalItemsCount / itemsPerPage)
-    } else {
-      // If currentPage or itemsPerPage aren't specified (or are invalid), issue the SQL query to
-      // get all listings (no pagination).
-      listings = await qb.getMany()
-
-      currentPage = 1
-      totalPages = 1
-      itemCount = listings.length
-      itemsPerPage = listings.length
-      totalItemsCount = listings.length
+    const paginate =
+      // currentPage and itemsPerPage are read in from the querystring, so we
+      // we confirm the type before proceeding
+      typeof paginationInfo.currentPage === "number" &&
+      paginationInfo.currentPage > 0 &&
+      typeof paginationInfo.itemsPerPage === "number" &&
+      paginationInfo.itemsPerPage > 0
+    if (paginate) {
+      // Calculate the number of listings to skip (because they belong to lower page numbers).
+      const offset = (paginationInfo.currentPage - 1) * paginationInfo.itemsPerPage
+      // Add the limit and offset to the inner query, so we only do the full
+      // join on the listings we want to show.
+      innerFilteredQuery.offset(offset).limit(paginationInfo.itemsPerPage)
     }
 
-    // Sort units and preferences.
-    // This step was removed from the SQL query because it interferes with pagination
-    // (See https://github.com/CityOfDetroit/affordable-housing-app/issues/88#issuecomment-865329223)
-    listings.forEach((listing) => {
-      listing.property.units.sort((a, b) => a.maxOccupancy - b.maxOccupancy)
-      listing.preferences.sort((a, b) => a.ordinal - b.ordinal)
-    })
-
-    /**
-     * Get the application counts and map them to listings
-     */
+    qb.andWhere("listings.id IN (" + innerFilteredQuery.getQuery() + ")")
+      // We set the inner WHERE params on the outer query, due to a bug in
+      // TypeORM: The WHERE params are dropped from the inner query.
```

Comment:
Do I understand correctly: because the WHERE params are dropped in the inner query, the inner query will return all listing IDs?

Reply:
I updated the comment to clarify, but the where params are just the strings that get escaped and inserted into the WHERE clause to replace the `:paramName` placeholders. We get a sql error when these are dropped because then the `:paramName` is left as-is instead of being replaced with something like "Fox Creek"

Reply:
They're not the full where clause, so the inner query should only be returning the listing IDs for the specific page of filtered listings that we're requesting.

Reply:
Ah I see. I was curious about this and so looked around for a bit and found https://github.com/typeorm/typeorm/blob/master/docs/select-query-builder.md#using-subqueries, where it seems like they're following a similar pattern. What they do:

```javascript
const userQb = await connection.getRepository(User)
    .createQueryBuilder("user")
    .select("user.name")
    .where("user.registered = :registered", { registered: true });

const posts = await connection.getRepository(Post)
    .createQueryBuilder("post")
    .where("post.title IN (" + userQb.getQuery() + ")")
    .setParameters(userQb.getParameters())
    .getMany();
```

Which makes me wonder if we could avoid `addFilters` having to pass back the parameters, and instead we could use `innerFilteredQuery.getParameters()`?

Reply:
Oh that's way better!